### PR TITLE
[Feature/extensions] Add getSettings support for AD

### PR DESCRIPTION
**Pull request by dbwiddis**
Companion PR: https://github.com/opensearch-project/opensearch-sdk-java/pull/147

NOTE: Merging this PR will break `main` on SDK project until the companion PR is merged.  Merge the SDK PR shortly after merging this one.

### Description

Registers custom Settings from an extension as dynamic settings in the Settings Module.

The pattern is similar to registering REST actions, with the complication that a `Setting<T>` needs to be passed over transport, and due to type erasure, `<T>` cannot be determined at runtime.  A `WriteableSetting` class determines this type by examining the default value and implements it for all types expected in the AD extension.

There will be a need to handle additional types in the future, such as List, and to track parsers and validators. This is tracked in [#148](https://github.com/opensearch-project/opensearch-sdk-java/issues/148)

### Issues Resolved

Fixes [#79](https://github.com/opensearch-project/opensearch-sdk-java/issues/79)

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
This is ready for review.  Notes:
 - Gradle Check is failing on BWC probably associated with #4513 
 - There is still more work to be done on SDK side to complete the linked issue, but that should not hold up merging this
 - It may be preferable to merge this earlier to unblock @joshpalis work on #4517 which needs some classes in this code and touches many of the same methods.

CC: @saratvemulapalli @owaiskazi19

File: server/src/test/java/org/opensearch/extensions/ExtensionResponseTests.java
Comment (ryanbogan) [inline server/src/test/java/org/opensearch/extensions/ExtensionResponseTests.java:20]: Small typo - booleanResponse

File: server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java
Comment (joshpalis) [inline server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java:46]: Are you planning on appending the unique ID of the extension node to the setting key prior to registration?

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java:46]: Yes, but it's complicated. Index scoped settings must begin with `index.` for example.  I haven't had time to look into all of those.  I think we need a new issue for this.

File: server/src/test/java/org/opensearch/extensions/ExtensionResponseTests.java
Comment (dbwiddis) [reply server/src/test/java/org/opensearch/extensions/ExtensionResponseTests.java:20]: Due to supply chain constraints, there's a vowel shortage.

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:110]: Can we log the default values here to provide more detail?

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:112]: Can we log the default types here and print something like:
`The current types supported are:.....`

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:136]: How about `genericType` -> `type`?

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:140]: `propArray` -> `propertyArray`

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:143]: Instead of having the switch case here and to make the code a little cleaner, can we create a map something like:
`{int: Setting.intSetting(params)}` and then fetch the value from the map?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:77]: Can we have the action in the format `internal:extensions...`

File: server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java:23]: `settings` -> `setting` and also the class name should be `RegisterSettingRequest`

File: server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java:24]: `SettingRequestHandler` here

**Comment by owaiskazi19** [comment]
Did initial pass. @dbwiddis and @saratvemulapalli do you think we should commit `WriteableSetting` to `main`?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:77]: This is an open question in #4334.  For now I'd like to leave it the same as the others.

File: server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java:23]: But it's a `List<Setting<?>>` and the method that kicks it off is `getSettings()`.  I'm okay with adding a word like "custom" here, but it's most definitely plural.

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:143]: Great idea, but I'm not sure it will work.
 - I'll need two maps, one for the fallbacks and one without (see the test class where I did exactly this)
   - And then I'd need to add the logic to test fallback first to determine which map to use
     - Except for the case with no fallback
 - In the future I expect a NamedWriteable type which will be a superclass
   - So I'd have to put in logic to check `instanceof` before looking it up... seems handling in `default` is easier here
 - I don't expect this list to grow very much and it seems to strike a good balance to me.

File: server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/settings/SettingsRequestHandler.java:24]: Similar to above. I clarified class javadoc with the words "list of custom extension settings".

**Comment by dbwiddis** [comment]
> Did initial pass. @dbwiddis and @saratvemulapalli do you think we should commit `WriteableSetting` to `main`?

I put in the minimum to handle the settings for AD, but there will be more work required in that class for the more general cases.  See [#148](https://github.com/opensearch-project/opensearch-sdk-java/issues/148).

File: server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java]: This makes sense. Thinking out loud: we need a way to differentiate `setting` vs `settings` as `settings` are already available for reading/writing from a stream.

File: server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java
Comment (owaiskazi19) [inline server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java:38]: Nit: Do we need another `map` here? Can we add a differentiator in the key?

File: server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java
Comment (owaiskazi19) [inline server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java:463]: Great to see tests for all the types

File: server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java]: I agree, we're really overloading the term!  Ultimately they are similar to Properties (with more functionality) so I'm fine with Settings as the class name but we do need to use another term for variable naming and javadocs.

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:111]: Missed this comment on my original review.  I actually don't think this can happen with the existing implementations due to the validators.  This is mostly a defensive catch against someone changing the code in the future, in which case I think the information given and the stack trace should provide enough detail.

File: server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java
Comment (dbwiddis) [reply server/src/test/java/org/opensearch/common/settings/WriteableSettingTests.java:38]: > Can we add a differentiator in the key for fallback? Feel free to ignore

It's an EnumMap so no.... unless we switch to a String-based map or make they key a `Pair<SettingType, Boolean>` ... except for the `Version` type it wouldn't need the boolean ...

Also we test them separately so I think the existing setup works.

So I will "feel free to ignore" although it's a great question. :)

**Review by joshpalis — APPROVED** [review]

File: server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/extensions/settings/RegisterSettingsRequest.java]: Done.

**Review by saratvemulapalli — APPROVED** [review]
Thanks @dbwiddis , I did a pass most of the changes look good to me.

**Pull request merged**
